rustls/src/server/hs.rs

```diff
@@ -120,6 +120,17 @@ impl ExtensionProcessing {
         #[cfg(feature = "quic")]
         {
             if cx.common.is_quic() {
+                // QUIC has strict ALPN, unlike TLS's more backwards-compatible behavior. RFC 9001
+                // says: The server MUST treat the inability to select a compatible application
+                // protocol as a connection error of type 0x0178
+                if cx.common.alpn_protocol.is_none()
+                    && (!our_protocols.is_empty() || maybe_their_protocols.is_some())
```

Comment:
This could similarly stand some prose that elaborates on what it's trying to achieve exactly, or maybe more exactly, why this is a good way to achieve what we're trying to achieve.

Reply:
Expanded on the rationale.